### DISCUSSION TITLE
Does this program convert UDP to TCP?

### DISCUSSION THREAD

**Discussion by jpmvferreira** [database64128/swgp-go#43] [Q&A]
I'm interested in establishing a Wireguard connection when inside a very restrictive firewall, which I believe blocks UDP traffic, and also has deep package inspection, although I haven't checked thoroughly.

Therefore, I would like to ask you:
1. Does this project convert UDP to TCP?
2. In an issue on this repository you mentioned that converting UDP to TCP is a bad idea. What would you suggest be done to bypass firewalls with these kind of restrictions?
3. Does Shadowsocks convert UDP to TCP?

To elaborate a little bit further on what I'm looking for, I am interested in using Shadowsocks to tunnel my Wireguard traffic. The reasons for that are:
1. My server's firewall is completely open to authenticated WG machines
2. Each device has an IP attributed by WG which I know by heart
4. I could tunnel all of my traffic via WG, creating a full VPN home, which would be obfuscated by Shadowsocks, making it very secure and private

If neither this program nor Shadowsocks converts UDP to TCP (or I end up going for Shadowsocks even though it does not perform that conversion), I can always chain this another program that does it, and have something that looks like:
*Client*: Wireguard -> UDP2TCP -> Shadowsocks -> Internet
*Server*: Internet -> TCP2UDP -> Wireguard
So far, I've found two programs that do this (not sure if there is a universally accepted recommended one):
- https://github.com/mullvad/udp-over-tcp
- https://github.com/wangyu-/udp2raw

**Comment by database64128** [comment]
https://github.com/SagerNet/sing-box implements a generic UoT encapsulation that works with most of its supported proxy protocols. You could use it with a tproxy or tun inbound to proxy all traffic.

**Reply by jpmvferreira** [reply]
Thanks for the reply. Do you happen to know how would it compare against chaining Shadowsocks with something like UDP2TCP? This software seems to be doing more than what I wanted to, and the UDP to TCP protocol is proprietary, which I would like to avoid.
Also, could I assume from your answer that this project does not support UDP to TCP?
